File: java/L100/LeetCode0143.java
package L100;

import util.ListNode;

public class LeetCode0143 {
    public void reorderList(ListNode head) {
        if (head == null || head.next == null || head.next.next == null) {
            return;
        }
        ListNode slow = head, fast = head;
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }
        ListNode ho = new ListNode(0);
        fast = slow.next;
        slow.next = null;
        while (fast != null) {
            ListNode t = fast.next;
            fast.next = ho.next;
            ho.next = fast;
            fast = t;
        }
        slow = head;
        ho = ho.next;
        while (slow != null && ho != null) {
            ListNode t1 = slow.next;
            ListNode t2 = ho.next;
            ho.next = slow.next;
            slow.next = ho;
            slow = t1;
            ho = t2;
        }
    }
}

File: cpp/L100/LeetCode0143.cc
#include "../heads/list_node.h"

class Solution {
    ListNode* reverse(ListNode *head) {
        ListNode dummy(0);
        while (head != nullptr) {
            ListNode *next = head->next;
            head->next = dummy.next;
            dummy.next = head;
            head = next;
        }
        return dummy.next;
    }
public:
    void reorderList(ListNode* head) {
        if (head == nullptr || head->next == nullptr) {
            return;
        }
        ListNode *slow{head}, *fast{head->next->next};
        while (fast != nullptr && fast->next != nullptr) {
            slow = slow->next;
            fast = fast->next->next;
        }
        ListNode *rList = reverse(slow->next);
        slow->next = nullptr;
        slow = head;
        while (rList != nullptr) {
            ListNode *next = rList->next;
            rList->next = slow->next;
            slow->next = rList;
            rList = next;
            if (slow->next->next != nullptr) { // 偶数情况
                slow = slow->next->next;
            } else { // 奇数情况：逆序链表多了一个数
                slow = slow->next;
            }
        }
    }
};